### PR TITLE
[dbnode] Enforce single Finalize on pooled objects

### PR DESCRIPTION
**Pull request by linasm**
**What this PR does / why we need it**:
Makes object pool panic on double `Finalize` of `RefCount`, tags `decoder` (which is not refcounted itself), or any other type that embeds `checked.FinalizeableOnce` struct.

**Special notes for your reviewer**:
A broader alternative to https://github.com/m3db/m3/pull/2603.

**Does this PR introduce a user-facing and/or backwards incompatible change?**:
NONE

**Does this PR require updating code package or user-facing documentation?**:
NONE

### CODE REVIEW

File: src/x/checked/finalizeable_once.go [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+}
+
+// Finalized returns true iff the object has been finalized.
+func (c *FinalizeableOnce) Finalized() bool {
+	return c.finalized.Load()
+}
+
+// SetFinalized sets the flag of object finalize.
+func (c *FinalizeableOnce) SetFinalized(f bool) {
+	c.finalized.Store(f)
```
Comment (robskillington) [inline src/x/checked/finalizeable_once.go:17]: I would use the underlying "Swap(value)" here and return what the old value was.

That way we can atomically see if we transitioned from false to true.

i.e. 
```go
func (c *FinalizeableOnce) SetFinalized() error {
  if oldValue := c.finalized.Swap(true); oldValue == true {
    return errAlreadyFinalized
  }
  return nil
}

func (c *FinalizeableOnce) SetUnfinalized() error {
  if oldValue := c.finalized.Swap(false); oldValue == false {
    return errNotFinalized
  }
  return nil
}
```

File: src/x/pool/object.go [hunk 2]
```diff
@@ -138,6 +148,15 @@ func (p *objectPool) Put(obj interface{}) {
 		p.onPoolAccessErrorFn(errPoolAccessBeforeInitialized)
 		return
 	}
+
+	if obj, ok := obj.(FinalizeableOnce); ok {
+		if obj.Finalized() {
+			p.onPoolAccessErrorFn(fmt.Errorf("double finalize, %v", obj))
```
Comment (robskillington) [inline src/x/pool/object.go:154]: nit: `double finalize, %v` -> `double finalize: %v` as per other errors using colon between message and value.

Comment (robskillington) [reply src/x/pool/object.go:154]: I would use `reflect.TypeOf(obj).String()` here. "%v" on `obj` will be fairly non-descript/not very useful.

File: src/x/checked/finalizeable_once.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package checked
+
+import "go.uber.org/atomic"
+
+// FinalizeableOnce implements the logics needed to enforce single Finalize on pool.ObjectPool.
+type FinalizeableOnce struct {
```
Comment (robskillington) [inline src/x/checked/finalizeable_once.go:6]: We should probably embed this to `RefCount` struct and get this checking for all things that embed `RefCount`.

File: src/x/checked/finalizeable_once.go [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package checked
```
Comment (robskillington) [inline src/x/checked/finalizeable_once.go:1]: Need license.

File: src/x/checked/finalizeable_once_test.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package checked
```
Comment (robskillington) [inline src/x/checked/finalizeable_once_test.go:1]: Need license.

File: src/x/serialize/decoder.go [hunk 6]
```diff
@@ -35,6 +35,7 @@ var (
 )
 
 type decoder struct {
+	checked.FinalizeableOnce
```
Comment (robskillington) [inline src/x/serialize/decoder.go:38]: Perhaps embed in the serialize/encoder too?

File: src/x/checked/finalizeable_once.go (hunk 1 above)
Comment (linasm) [reply src/x/checked/finalizeable_once.go:17]: Nice idea, changing.

File: src/x/pool/object.go (hunk 2 above)
Comment (linasm) [reply src/x/pool/object.go:154]: 👍

File: src/x/checked/finalizeable_once.go (hunk 3 above)
Comment (linasm) [reply src/x/checked/finalizeable_once.go:6]: Already a part of this PR.

File: src/x/serialize/decoder.go (hunk 6 above)
Comment (linasm) [reply src/x/serialize/decoder.go:38]: Done.

**Comment by linasm** [comment]
@robskillington I've finished addressing your feedback on this for completeness, but I am putting this PR on hold (draft) for the time being, as we are still exploring alternative approaches.

**Pull request closed without merging**